pkg/chunked/storage_linux.go

```diff
@@ -1280,6 +1280,18 @@ func typeToOsMode(typ string) (os.FileMode, error) {
 	return 0, fmt.Errorf("unknown file type %q", typ)
 }
 
+func isRootDirectory(path string) bool {
```

Comment:
I’m rather unhappy about this extra and somewhat unusual code path. Could we have `safeMkdir` return a “resolved to `dirfd`” value which precisely matches its decision?

Reply:
we don't need this for safeMkdir but to detect what tar entry refers to the root, so we can store its file mode and report it to the caller

Reply:
Yes — but see https://github.com/containers/storage/pull/2194/files#r1876699526 . I think all of the root, and root`/..` , handling, needs _some_ kind of cleanup / restructuring, and I suspect that would be fairly likely to make this function unnecessary.

My first instinct would be to factor out the repeated 
```go
	parent, base := filepath.Split(name)
	parentFd := dirfd
	if parent != "" && parent != "." {
		parentFile, err := openOrCreateDirUnderRoot(dirfd, parent, 0)
		if err != nil {
			return err
		}
		defer parentFile.Close()
		parentFd = int(parentFile.Fd())
	}
```
into a single helper (preparePathForCreation returning an object?). And then, I don’t know, _maybe_ call that in `ApplyDiff`, and centralize that way?

I haven’t given this much thought now.

Reply:
`cleanAbsDirectory` + a single comparison to `"/"` is nice, short and elegant.